gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/GCImpl.java

```diff
@@ -0,0 +1,202 @@
+ * all named-references.
+ */
+final class GCImpl implements GC {
+  private static final Logger LOGGER = LoggerFactory.getLogger(GCImpl.class);
+
+  private final NessieApiV1 api;
+  private final Function<Reference, Instant> cutOffTimestamp;
+
+  GCImpl(NessieApiV1 api, Function<Reference, Instant> cutOffTimestamp) {
+    this.api = api;
+    this.cutOffTimestamp = cutOffTimestamp;
+  }
+
+  @Override
+  public <GC_CONTENT_VALUES extends ExpiredContentValues> GCResult<GC_CONTENT_VALUES> performGC(
+      ContentValuesCollector<GC_CONTENT_VALUES> contentValuesCollector) {
+    // Algorithm for performing the GC to identify the expired contents :
+
+    // Walk through each reference.
+    // For each commit in a reference, collect live table keys or dropped table keys.
+    // For each key, get the content.
+    // If the content's key is in dropped table keys,
+    //     add content to expired set (If the content is not live from previous walks of other
+    // references).
+    // Else if the commit is head commit, add contents to live set.
+    // Else check if the content is expired or not based on cutoff time of this reference.
+    //     Based on that add content to respective sets (live or expired).
+    //
+    // Note: So, If a table is dropped in a reference. Then independent of cutoff time,
+    // all the snapshots for that table in the reference (that are not live in other reference)
+    // will be considered for expiry.
+
+    List<Reference> references = api.getAllReferences().get().getReferences();
+    for (Reference reference : references) {
+      Instant cutOffTimestamp = this.cutOffTimestamp.apply(reference);
+      walkReference(
+          contentValuesCollector,
+          reference,
+          commitMeta ->
+              // If the commit time is older than (think: less than, before) cutoff-time, then
+              // commit is expired.
+              Objects.requireNonNull(commitMeta.getCommitTime()).compareTo(cutOffTimestamp) < 0);
+    }
+    return new GCResult<>(contentValuesCollector.contentValues);
+  }
+
+  private <GC_CONTENT_VALUES extends ExpiredContentValues> void walkReference(
+      ContentValuesCollector<GC_CONTENT_VALUES> contentValuesCollector,
+      Reference reference,
+      Predicate<CommitMeta> expiredCommitPredicate) {
+    try (Stream<LogResponse.LogEntry> commits =
+        StreamingUtil.getCommitLogStream(
+            api, reference.getName(), null, null, null, OptionalInt.empty(), true)) {
+
+      Set<ContentKey> deletedKeys = new HashSet<>();
+      AtomicBoolean commitHead = new AtomicBoolean(true);
+      commits.forEach(
+          (logEntry) -> {
+            boolean isExpired;
+            if (commitHead.get()) {
+              // If the commit is head commit, then it should be kept live if the table is reachable
+              isExpired = false;
+              commitHead.set(false);
```

Comment:
The `Stream.forEach` javadoc explicitly declares that the order of callbacks may not match the natural order of elements in the stream. I believe current logic of determining the head commit is quite risky and error-prone.

Reply:
I think stream.paralell().foreach cannot guarantee the order. But the stream.foreach() will give in original order.
But yeah safe to use forEachOrdered instead of forEach